### ISSUE TITLE
Data science library 'kernel spec' is not installed in interpreter Python 3.8.2 (64-bit) 

### ISSUE THREAD

**Issue by ramdop** [microsoft/vscode-jupyter#1056]
# Bug: Notebook Editor, Interactive Window, Editor cells

## Steps to cause the bug to occur

1. VS Code V1.45 and Python 3.8.2
2. Python extension is installed
3. Right-click on a python file and click on "Run current file in an interactive window"

## Actual behavior
![image](https://user-images.githubusercontent.com/13087620/81817266-19ba9c00-954a-11ea-84fa-34f9c5793ba9.png)
Doesn't start the Jupyter notebook
## Expected behavior
File runs in an interactiv jupyter notebook. 

### Your Jupyter and/or Python environment

_Please provide as much info as you readily know_

-   **Jupyter server running:** not started | No Kernel : Not started
-   **Extension version:** 2020.05.78807
-   **VS Code version:** 1.45
-   **Setting python.jediEnabled:** true | false
-   **Python and/or Anaconda version: ** Python 3.8.2 64bit
-   **OS:** Windows
-   **Virtual environment:** conda | venv | virtualenv | N/A | ...

## Developer Tools Console Output

Microsoft Data Science for VS Code Engineering Team: @rchiodo, @IanMatthewHuff, @DavidKutu, @DonJayamanne, @greazer

**Comment by Kuligha** [comment]
I am currently experiencing this issue

**Comment by vtrikoupis** [comment]
Same issue on Code - Insiders

**Comment by IanMatthewHuff** [comment]
I believe that this is different from the original issue. But reactivating to make sure that we are tracking this.

**Comment by IanMatthewHuff** [comment]
@DonJayamanne Looks like this might be the install issue that you fixed recently?

**Comment by rchiodo** [comment]
On VS code insiders this should be fixed in our next release.

**Comment by DevilXD** [comment]
I've just ran into this issue myself. The error log mentions something worrying, about a circular import in Cython?

```
Error 2021-04-02 19:51:41: Failed to get a list of the kernelspec using python script [i [Error]: Traceback (most recent call last):
  File "c:\Users\DevilXD\.vscode\extensions\ms-toolsai.jupyter-2021.5.702919634\pythonFiles\vscode_datascience_helpers\getJupyterKernelspecVersion.py", line 3, in <module>
    import jupyter_client
  File "C:\Users\DevilXD\AppData\Roaming\Python\Python38\site-packages\jupyter_client\__init__.py", line 4, in <module>
    from .connect import *
  File "C:\Users\DevilXD\AppData\Roaming\Python\Python38\site-packages\jupyter_client\connect.py", line 21, in <module>
    import zmq
  File "C:\Users\DevilXD\AppData\Roaming\Python\Python38\site-packages\zmq\__init__.py", line 47, in <module>
    from zmq import backend
  File "C:\Users\DevilXD\AppData\Roaming\Python\Python38\site-packages\zmq\backend\__init__.py", line 40, in <module>
    reraise(*exc_info)
  File "C:\Users\DevilXD\AppData\Roaming\Python\Python38\site-packages\zmq\utils\sixcerpt.py", line 34, in reraise
    raise value
  File "C:\Users\DevilXD\AppData\Roaming\Python\Python38\site-packages\zmq\backend\__init__.py", line 27, in <module>
    _ns = select_backend(first)
  File "C:\Users\DevilXD\AppData\Roaming\Python\Python38\site-packages\zmq\backend\select.py", line 28, in select_backend
    mod = __import__(name, fromlist=public_api)
  File "C:\Users\DevilXD\AppData\Roaming\Python\Python38\site-packages\zmq\backend\cython\__init__.py", line 6, in <module>
    from . import (constants, error, message, context,
ImportError: cannot import name 'constants' from partially initialized module 'zmq.backend.cython' (most likely due to a circular import) (C:\Users\DevilXD\AppData\Roaming\Python\Python38\site-packages\zmq\backend\cython\__init__.py)

	at ChildProcess.<anonymous> (c:\Users\DevilXD\.vscode\extensions\ms-toolsai.jupyter-2021.5.702919634\out\client\extension.js:24:222480)
	at Object.onceWrapper (events.js:422:26)
	at ChildProcess.emit (events.js:315:20)
	at maybeClose (internal/child_process.js:1021:16)
	at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)] {
  category: 'unknown'
}
Error 2021-04-02 19:51:41: Exception attempting to enumerate kernelspecs:  [i [Error]: Traceback (most recent call last):
  File "c:\Users\DevilXD\.vscode\extensions\ms-toolsai.jupyter-2021.5.702919634\pythonFiles\pyvsc-run-isolated.py", line 32, in <module>
    runpy.run_module(module, run_name="__main__", alter_sys=True)
  File "C:\Python38\lib\runpy.py", line 207, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "C:\Python38\lib\runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "C:\Python38\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\DevilXD\AppData\Roaming\Python\Python38\site-packages\jupyter.py", line 10, in <module>
    run_module('jupyter_core')
  File "C:\Python38\lib\runpy.py", line 210, in run_module
    return _run_code(code, {}, init_globals, run_name, mod_spec)
  File "C:\Python38\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\DevilXD\AppData\Roaming\Python\Python38\site-packages\jupyter_core\__main__.py", line 3, in <module>
    main()
  File "C:\Users\DevilXD\AppData\Roaming\Python\Python38\site-packages\jupyter_core\command.py", line 247, in main
    command = _jupyter_abspath(subcommand)
  File "C:\Users\DevilXD\AppData\Roaming\Python\Python38\site-packages\jupyter_core\command.py", line 133, in _jupyter_abspath
    raise Exception(
Exception: Jupyter command `jupyter-kernelspec` not found.

	at ChildProcess.<anonymous> (c:\Users\DevilXD\.vscode\extensions\ms-toolsai.jupyter-2021.5.702919634\out\client\extension.js:24:222480)
	at Object.onceWrapper (events.js:422:26)
	at ChildProcess.emit (events.js:315:20)
	at maybeClose (internal/child_process.js:1021:16)
	at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)] {
  category: 'unknown'
}
Error 2021-04-02 19:51:41: Kernel spec not found:  [i [Error]: Traceback (most recent call last):
  File "c:\Users\DevilXD\.vscode\extensions\ms-toolsai.jupyter-2021.5.702919634\pythonFiles\pyvsc-run-isolated.py", line 32, in <module>
    runpy.run_module(module, run_name="__main__", alter_sys=True)
  File "C:\Python38\lib\runpy.py", line 207, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "C:\Python38\lib\runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "C:\Python38\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\DevilXD\AppData\Roaming\Python\Python38\site-packages\jupyter.py", line 10, in <module>
    run_module('jupyter_core')
  File "C:\Python38\lib\runpy.py", line 210, in run_module
    return _run_code(code, {}, init_globals, run_name, mod_spec)
  File "C:\Python38\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\DevilXD\AppData\Roaming\Python\Python38\site-packages\jupyter_core\__main__.py", line 3, in <module>
    main()
  File "C:\Users\DevilXD\AppData\Roaming\Python\Python38\site-packages\jupyter_core\command.py", line 247, in main
    command = _jupyter_abspath(subcommand)
  File "C:\Users\DevilXD\AppData\Roaming\Python\Python38\site-packages\jupyter_core\command.py", line 133, in _jupyter_abspath
    raise Exception(
Exception: Jupyter command `jupyter-kernelspec` not found.

	at ChildProcess.<anonymous> (c:\Users\DevilXD\.vscode\extensions\ms-toolsai.jupyter-2021.5.702919634\out\client\extension.js:24:222480)
	at Object.onceWrapper (events.js:422:26)
	at ChildProcess.emit (events.js:315:20)
	at maybeClose (internal/child_process.js:1021:16)
	at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)] {
  category: 'unknown',
  vslsStack: [ CallSite {}, CallSite {}, CallSite {}, CallSite {}, CallSite {} ]
}
Info 2021-04-02 19:51:41: Using colors to generate CSS ...
Info 2021-04-02 19:51:41: Using colors to generate CSS ...
Info 2021-04-02 19:51:41: Using colors to generate CSS ...
Info 2021-04-02 19:51:41: Using colors to generate CSS ...
Info 2021-04-02 19:51:41: Using colors to generate CSS ...
Info 2021-04-02 19:51:41: Using colors to generate CSS ...
Info 2021-04-02 19:51:41: Using colors to generate CSS ...
Error 2021-04-02 19:51:48: Install Missing Dependencies, Class name = v, completed in 12336ms, has a falsy return value, Return Value: undefined [a [Error]: IPyKernel not installed into interpreter Python 3.8.8 64-bit:C:\Python38\python.exe
	at v.installMissingDependencies (c:\Users\DevilXD\.vscode\extensions\ms-toolsai.jupyter-2021.5.702919634\out\client\extension.js:49:561216)
	at processTicksAndRejections (internal/process/task_queues.js:97:5)] {
  category: 'noipykernel',
  reason: 2
}
Error 2021-04-02 19:51:48: Raw session failed to start because dependencies not installed
Info 2021-04-02 19:51:48: Shutdown session -- complete
Error 2021-04-02 19:51:48: Failed to determine if a notebook is active for the current editor [a [Error]: IPyKernel not installed into interpreter Python 3.8.8 64-bit:C:\Python38\python.exe
	at v.installMissingDependencies (c:\Users\DevilXD\.vscode\extensions\ms-toolsai.jupyter-2021.5.702919634\out\client\extension.js:49:561216)
	at processTicksAndRejections (internal/process/task_queues.js:97:5)] {
  category: 'noipykernel',
  reason: 2,
  vslsStack: [ CallSite {}, CallSite {} ]
}
Error 2021-04-02 19:51:48: DataScience Error [a [Error]: IPyKernel not installed into interpreter Python 3.8.8 64-bit:C:\Python38\python.exe
	at v.installMissingDependencies (c:\Users\DevilXD\.vscode\extensions\ms-toolsai.jupyter-2021.5.702919634\out\client\extension.js:49:561216)
	at processTicksAndRejections (internal/process/task_queues.js:97:5)] {
  category: 'noipykernel',
  reason: 2,
  vslsStack: [ CallSite {}, CallSite {} ]
}
```

The result of this is the same as the issue describes, I keep telling it to Install it, and it says that everything is satisfied. The Jupyter server refuses to start, even after selecting it, and reinstalling all extensions.

**Comment by DonJayamanne** [comment]
@DevilXD Please could you file a seprate issue

**Comment by nevr1106** [comment]
@DevilXD The key issue is that you have to have jupyter kernal in the python environment you are running in. The jupyter extension does not do that.
For me, the best practice is ensuring I install jupyterlab or jupyter notebook when I create an environment. For example, this is how I create my environment:
1. start windows terminal with a connection to conda.
2. conda create -n dummy_env python **notebook** pandas numpy
3. open vscode from windows or directly from comand line with `code .`
4. select dummy_env
5. run notebook in vscode without a issue

The jupyter extension seems to give people an impression that the notebook package is not required in the running environment, which is not the case.

**Comment by rchiodo** [comment]
> The jupyter extension seems to give people an impression that the notebook package is not required in the running environment, which is not the case.

This is not true. Ipykernel is the thing we require. And we will install it into the kernel you're using.  

The error @DevilXD  is having is likely related to this:
https://github.com/jupyter/notebook/issues/5709

That error is preventing our install from working (or we're installing the wrong ipykernel)

**Comment by nevr1106** [comment]
When I had this issue awhile ago. vscode does prompt a notice saying ipykernel needs to be installed. 

However, when I click the install button, the issue persists. So I agree that there might be some issues with the installation of ipykernel within vscode.

If the kernel is installed in the environment, this issue can be avoided.

**Comment by DevilXD** [comment]
@nevr1106 The thing is, I'm not using any environments, as far as I'm aware. There's no venvs in play here. I'm not using conda, pandas or numpy either, I just tried to show a friend how to use the jupyter notepad, that I'm normally not using.

Reinstalling `pyzmq` (from `pyzmq-19.0.1` to `pyzmq-22.0.3`, apparently) gave me an `ImportError: DLL load failed while importing win32api: %1 is not a valid Win32 application` instead, and the kernel still refused to start. So I googled the error, and found this: https://stackoverflow.com/questions/19019720/importerror-dll-load-failed-1-is-not-a-valid-win32-application-but-the-dlls

Downloading and installing `pywin32-300-cp38-cp38-win_amd64` appears to have fixed the issue (also the version changed from 228 to 300): https://i.imgur.com/mMpoEDA.png

I have no idea why all of this happened, but I remember trying out the jupyter notebooks a while ago (about a year), and everything worked fine back then, when I clicked on the `Install` button. Just saying.
